### PR TITLE
Authorization app owners

### PR DESCRIPTION
**Pull request by harshith-venkatesh**
PR for Authorize the user for post and update of the tasks #141 new PR

### CODE REVIEW

**Comment by Harshi7016** [comment]
Sorry for too many commits , I will be raising a new PR once the tests get passed

File: controllers/usersController.js [hunk 1]
```diff
@@ -22,6 +23,27 @@ const getUsers = async (req, res) => {
...
+/**
+ * Fetches the data about our account Owners
+ *
+ * @param req {Object} - Express request object
+ * @param res {Object} - Express response object
+ */
+
+const getAccountOwners = async (req, res) => {
+  try {
+    // const allUsers = await userQuery.fetchUsers(req.query)
```
Comment (Harshi7016) [inline controllers/usersController.js:35]: remove comments

File: controllers/usersController.js [hunk 2]
```diff
@@ -22,6 +23,27 @@ const getUsers = async (req, res) => {
...
+    // return res.json({
+    //   message: 'Users returned successfully!',
+    //   users: allUsers,
+    // })
+    return accountOwners
+  } catch (error) {
+    logger.error(`Error while fetching application owners: ${error}`)
+    return res.boom.serverUnavailable('Something went wrong please contact admin')
```
Comment (Harshi7016) [inline controllers/usersController.js:43]: error message: badImple- 500

File: middlewares/authorizeOwner.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
+const usersController = require('../controllers/usersController')
+/**
+ * Middleware to validate the authorized routes to be able to create & Update tasks
+ * 1] Verifies the user's role as Application owner
+ * * 2] In case of absence of user role, error is invoked
+ *
+ * The currently implemented mechanism satisfies the current use case.
+ * Authentication with JWT and a refreshToken to be added once we have user permissions and authorizations to be handled
+ *
+ * @todo: Add tests to assert on refreshed JWT generation by modifying the TTL values for the specific test. Currently not possible in the absence of a test-suite.
```
Comment (Harshi7016) [inline middlewares/authorizeOwner.js:10]: remove todo

File: middlewares/authorizeOwner.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+module.exports = async (req, res, next) => {
+  try {
+    // get user data from `req.userData` for further use
+    const accountOwners = await usersController.getAccountOwners()
+    const { username } = req.userData
+
+    if (accountOwners.filter((owner) => owner.username === username)) {
+      return next()
+    } else {
+      return res.boom.notFound('Unauthorized User')
```
Comment (Harshi7016) [inline middlewares/authorizeOwner.js:27]: 403

File: middlewares/authorizeOwner.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    }
+  } catch (err) {
+    logger.error(err)
+    return res.boom.badImplementation('Unauthorized User')
```
Comment (Harshi7016) [inline middlewares/authorizeOwner.js:31]: check other middleware

File: test/integration/tasksController.test.js [hunk 6]
```diff
@@ -5,10 +5,12 @@ const chaiHttp = require('chai-http')
 
 const app = require('../../server')
 const tasks = require('../../models/tasks')
-
+const addUser = require('../utils/addUser')
+const authService = require('../../services/authService')
 chai.use(chaiHttp)
 
 describe('Tasks', function () {
+  let jwt = ''
```
Comment (Harshi7016) [inline test/integration/tasksController.test.js:13]: let jwt;

File: test/integration/tasksController.test.js [hunk 7]
```diff
@@ -5,10 +5,12 @@ const chaiHttp = require('chai-http')
...
   let tid = ''
```
Comment (Harshi7016) [inline test/integration/tasksController.test.js:14]: let tid;

File: test/integration/tasksController.test.js [hunk 8]
```diff
@@ -118,6 +124,7 @@ describe('Tasks', function () {
       chai
         .request(app)
         .patch('/tasks/taskid')
+        .set('cookie', `rds-session=${jwt}`)
```
Comment (Harshi7016) [inline test/integration/tasksController.test.js:127]: create mock for forbidden with other user details

**Comment by ankurnarkhede** [comment]
CI failing... @harshith-venkatesh

**Review by ankurnarkhede — CHANGES_REQUESTED** [review]
CI failing...

**Comment by harshith-venkatesh** [comment]
@ankurnarkhede I will update the test cases today
Sorry for the delay

**Comment by ankurnarkhede** [comment]
@harshith-venkatesh , this PR is pending since long. Can you check it?

**Comment by harshith-venkatesh** [comment]
@ankurnarkhede , I am finding issue in creating test cases, Can I send PR without 403 error test case?

**Comment by harshith-venkatesh** [comment]
Closing this PR on discussion with @ankurnarkhede and creating a new PR with CI being successful
@ankushdharkar

**Pull request closed without merging**